packages/client/src/authentication/adapters/Credentials.ts

```diff
@@ -0,0 +1,38 @@
+import { DateUtils, invariant } from '@lens-protocol/shared-kernel';
+import jwtDecode, { JwtPayload } from 'jwt-decode';
+
+// Threshold in seconds that will mark token as expired even it's still valid
+// Adds some time for all communications that's required to refresh tokens
+const TOKEN_EXP_THRESHOLD = DateUtils.secondsToMs(3);
+
+export class Credentials {
+  constructor(readonly accessToken: string | undefined, readonly refreshToken: string) {}
+
+  canRefresh(): boolean {
+    const now = Date.now();
```

Comment:
Thinking more about this, how does this work if client clock is skewed?

Reply:
it won't work but I don't know how we can go around it

Reply:
**TL;DR;**
I would say let's compare clock time with `iat` timestamp when we get the credentials the first time (and at every renew time) and if the time discrepancy is too large we can throw a dedicated `ClockSkewedError` so the integrator can do something about it.

There are few strategies that can work but relies on assumptions on the number of processes/threads and the storage.

More or less all based on storing the client time at the same time of the credentials and use relative difference and the server knowledge of current time at the time the credentials has been issued (i.e. the `iat` claim in the Access Token JTW is a good one).

But it falls short if you consider the low-level client is designed to work on many scenarios/enviornments. From browser with local storage as credentials storage to a server side with multiple instances of the client process with a shared storage (e.g. think about a dockerized process with a shared redis as storage). In the latter all client process must have clock in sync, which probably is fine in most docker/containerized environments, but is an implicit assumption nonetheless.

Reply:
I added check if clock is not skewed, based on system time and token issue time (https://github.com/lens-protocol/lens-sdk/pull/148/commits/ec446db89d23f33bc88f00120a8ab19f7c1e031f)
but it makes tests harder, I skipped two tests for now, I guess I have to introduce mocked auth server response to have control over the credentials returned.

Reply:
I added `checkClock` method on Credentials class, it makes testing with useFakeTimers harder :)